internal/integration/unified/client_entity.go

```diff
@@ -326,8 +344,39 @@ func (c *clientEntity) failedEvents() []*event.CommandFailedEvent {
 	return events
 }
 
-func (c *clientEntity) poolEvents() []*event.PoolEvent {
-	return c.pooled
+// filterEventsBySeq filters events by sequence number for the given eventType.
+// See comments on eventSequencer for more details.
+func filterEventsBySeq[T any](c *clientEntity, events []T, eventType monitoringEventType) []T {
+	cutoff := c.eventSequencer.cutoff.Load()
+	if cutoff == 0 {
+		return events
+	}
+
+	// Lock order: eventProcessMu -> eventSequencer.mu (matches writers)
+	c.eventProcessMu.RLock()
+	c.eventSequencer.mu.RLock()
+
+	// Snapshot to minimize time under locks and avoid races
+	eventsSnapshot := slices.Clone(events)
+	seqSnapshot := slices.Clone(c.eventSequencer.seqByEventType[eventType])
+
+	c.eventSequencer.mu.RUnlock()
+	c.eventProcessMu.RUnlock()
```

Comment:
I assume we need to hold the `eventsProcessMu` read lock here to prevent race conditions reading `events`. That requirement is difficult to trace and may lead to bugs if future implementers don't realize the purpose (it's unconventional to require a function to take a lock to read an input param). I recommend taking a read lock before calling `filterEventsBySeq`.